sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaProtoTable.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.meta.provider.kafka;
+
+import com.google.protobuf.Message;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import org.apache.beam.sdk.extensions.protobuf.ProtoCoder;
+import org.apache.beam.sdk.extensions.protobuf.ProtoMessageSchema;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+public class BeamKafkaProtoTable<ProtoT extends Message> extends BeamKafkaTable {
+  private final transient Class<ProtoT> protoClass;
+
+  public BeamKafkaProtoTable(
+      Schema beamSchema, String bootstrapServers, List<String> topics, Class<ProtoT> protoClass) {
+    super(beamSchema, bootstrapServers, topics);
+    this.protoClass = protoClass;
```

Comment:
We should verify early that `beamSchema` and the schema inferred from `protoClass` by `ProtoSchemaRegistry` are equivalent. In theory we could also allow the table definition to omit the schema in the case, since it can be determined fully from the protoClass.

Reply:
Great idea. I removed beamSchema and infer the schema from the protoClass. It seems to work.

Reply:
Nice!

I think we should still handle the situation where a schema is specified though, and raise an error if the schema inferred from the proto class is not [equivalent](https://github.com/apache/beam/blob/7828af87e55a1311fcd0a76070713b4eac8b2c4c/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java#L343) to the one the user specified.

Reply:
Done. I also changed assignableTo to equivalent in toProtoFn. I've just learned that proto3 removed nulls.